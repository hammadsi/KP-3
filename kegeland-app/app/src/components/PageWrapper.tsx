import React from 'react';
import {StyleProp, StyleSheet, ViewStyle} from 'react-native';
import {Title} from 'react-native-paper';
import {SafeAreaView} from 'react-native-safe-area-context';

type ContentSize = 'small' | 'medium' | 'large' | 'full';

export type PageWrapperProps = {
  title: string;
  contentSize?: ContentSize;
  style?: StyleProp<ViewStyle>;
  testID?: string;
  children: React.ReactNode;
};

/**
 * Retrieves the page's padding based on it's defined content size
 * @param contentSize the content size
 * @see {@link ContentSize}
 * @returns the padding
 */
const getPadding = (contentSize?: ContentSize) => {
  switch (contentSize) {
    case 'full':
      return 0;
    case 'large':
      return 8;
    case 'medium':
      return 16;
    default:
      return 32;
  }
};

/**
 * Component for rendering page layout.
 * @param props the props
 * @see {@link PageWrapperProps}
 */
const PageWrapper: React.FC<PageWrapperProps> = ({
  title,
  contentSize,
  style,
  testID,
  children,
}) => {
  return (
    <SafeAreaView
      testID={testID}
      style={[
        styles.wrapper,
        {paddingHorizontal: getPadding(contentSize)},
        style,
      ]}>
      <Title style={styles.title}>{title}</Title>
      {children}
    </SafeAreaView>
  );
};

PageWrapper.defaultProps = {
  contentSize: 'small',
};

const styles = StyleSheet.create({
  wrapper: {
    height: '100%',
    width: '100%',
    marginTop: 8,
  },
  title: {
    fontWeight: 'bold',
    fontSize: 24,
  },
});

export default PageWrapper;
