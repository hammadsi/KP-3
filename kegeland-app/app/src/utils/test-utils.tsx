/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
