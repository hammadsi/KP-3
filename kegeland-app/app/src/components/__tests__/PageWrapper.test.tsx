import React from 'react';
import {Text} from 'react-native';

import {render} from '~utils/test-utils';

import PageWrapper, {PageWrapperProps} from '../PageWrapper';

describe('Test PageWrapper-component', () => {
  it('should render correctly', () => {
    const props: Omit<PageWrapperProps, 'children'> = {
      title: 'Test page',
    };
    const component = (
      <PageWrapper {...props}>
        <Text>Test</Text>
      </PageWrapper>
    );
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render correctly with content size = full', () => {
    const props: Omit<PageWrapperProps, 'children'> = {
      title: 'Test page',
      contentSize: 'full',
    };
    const component = (
      <PageWrapper {...props}>
        <Text>Test</Text>
      </PageWrapper>
    );
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render correctly with content size = large', () => {
    const props: Omit<PageWrapperProps, 'children'> = {
      title: 'Test page',
      contentSize: 'large',
    };
    const component = (
      <PageWrapper {...props}>
        <Text>Test</Text>
      </PageWrapper>
    );
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render correctly with content size = medium', () => {
    const props: Omit<PageWrapperProps, 'children'> = {
      title: 'Test page',
      contentSize: 'medium',
    };
    const component = (
      <PageWrapper {...props}>
        <Text>Test</Text>
      </PageWrapper>
    );
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
