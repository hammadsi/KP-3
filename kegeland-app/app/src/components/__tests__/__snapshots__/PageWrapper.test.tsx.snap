// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PageWrapper-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    edges={
      Array [
        "bottom",
        "left",
        "right",
        "top",
      ]
    }
    style={
      Array [
        Object {
          "height": "100%",
          "marginTop": 8,
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 32,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Array [
            Object {
              "textAlign": "left",
            },
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "500",
              "writingDirection": "ltr",
            },
            Array [
              Object {
                "fontSize": 20,
                "letterSpacing": 0.15,
                "lineHeight": 30,
                "marginVertical": 2,
              },
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              },
            ],
          ],
        ]
      }
    >
      Test page
    </Text>
    <Text>
      Test
    </Text>
  </RNCSafeAreaView>
</View>
`;

exports[`Test PageWrapper-component should render correctly with content size = full 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    edges={
      Array [
        "bottom",
        "left",
        "right",
        "top",
      ]
    }
    style={
      Array [
        Object {
          "height": "100%",
          "marginTop": 8,
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 0,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Array [
            Object {
              "textAlign": "left",
            },
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "500",
              "writingDirection": "ltr",
            },
            Array [
              Object {
                "fontSize": 20,
                "letterSpacing": 0.15,
                "lineHeight": 30,
                "marginVertical": 2,
              },
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              },
            ],
          ],
        ]
      }
    >
      Test page
    </Text>
    <Text>
      Test
    </Text>
  </RNCSafeAreaView>
</View>
`;

exports[`Test PageWrapper-component should render correctly with content size = large 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    edges={
      Array [
        "bottom",
        "left",
        "right",
        "top",
      ]
    }
    style={
      Array [
        Object {
          "height": "100%",
          "marginTop": 8,
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 8,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Array [
            Object {
              "textAlign": "left",
            },
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "500",
              "writingDirection": "ltr",
            },
            Array [
              Object {
                "fontSize": 20,
                "letterSpacing": 0.15,
                "lineHeight": 30,
                "marginVertical": 2,
              },
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              },
            ],
          ],
        ]
      }
    >
      Test page
    </Text>
    <Text>
      Test
    </Text>
  </RNCSafeAreaView>
</View>
`;

exports[`Test PageWrapper-component should render correctly with content size = medium 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    edges={
      Array [
        "bottom",
        "left",
        "right",
        "top",
      ]
    }
    style={
      Array [
        Object {
          "height": "100%",
          "marginTop": 8,
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 16,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Array [
            Object {
              "textAlign": "left",
            },
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "500",
              "writingDirection": "ltr",
            },
            Array [
              Object {
                "fontSize": 20,
                "letterSpacing": 0.15,
                "lineHeight": 30,
                "marginVertical": 2,
              },
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              },
            ],
          ],
        ]
      }
    >
      Test page
    </Text>
    <Text>
      Test
    </Text>
  </RNCSafeAreaView>
</View>
`;
